### PR TITLE
CICD pipeline and integration tests

### PR DESCRIPTION
**Pull request by majguo**
This is a follow-up PR for #4, which includes:

* Sample apps (Open Liberty and Quarkus) for integration test 
* CICD pipeline
* Add/update readme
* Add javadoc

## Testing

See the successful build workflow run which executed the unit tests and integration tests with Azure services:

* https://github.com/majguo/azure-microprofile/actions/runs/5998279469

Signed-off-by: Jianguo Ma <jiangma@microsoft.com>

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+jobs:
+  build:
+    permissions:
+      id-token: write
+    environment: ci
+    env:
+      AZURE_CLIENT_ID: ${{ vars.AZURE_CLIENT_ID }}
+      AZURE_TENANT_ID: ${{ vars.AZURE_TENANT_ID }}
+      AZURE_SUBSCRIPTION_ID: ${{ vars.AZURE_SUBSCRIPTION_ID }}
```
Comment (edburns) [inline .github/workflows/build.yml:41]: I request that you read these from secrets, as we have done in our other repositories, such as [liberty.aks](https://github.com/WASdev/azure.websphere-traditional.cluster/blob/f4103d4dd034174060b87a803efb5435355561bb/.github/workflows/integration-test.yaml#L64).

File: integration-tests/quarkus-sample/README.md [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+Edit the file `src/main/resources/META-INF/microprofile-config.properties`. Uncomment all the lines. Make it so the value of the config property `azure.keyvault.url` is the value of `$AZURE_KEYVAULT_URL`.
+
+## Running the sample
+
+Now you can launch the sample with:
+
+```
+mvn package quarkus:run
+```
+
```
Comment (edburns) [inline integration-tests/quarkus-sample/README.md:86]: I see this error on in the quarkus output. If this is expected, we should inform the reader that it is expected.

```
[ERROR] WorkloadIdentityCredential authentication unavailable. The workload options are not fully configured. See the troubleshooting guide for more information. https://aka.ms/azsdk/java/identity/workloadidentitycredential/troubleshoot
```

File: microprofile-config-keyvault/src/main/java/com/azure/microprofile/config/keyvault/NoCacheAzureKeyVaultOperation.java [hunk 3]
```diff
@@ -7,22 +7,47 @@
...
+ * This class is used to fetch the secrets from Azure Key Vault.
+ *
+ * <ul>
+ *     <li>It does not cache the secrets. It fetches the secrets from Azure Key Vault every time.</li>
+ *     <li>A secret name regular expression is used to filter out invalid secret names to reduce unnecessary calls to Azure Key Vault.</li>
+ *     <li>The default value of the secret name regular expression is "^[0-9a-zA-Z-]+$".</li>
+ * </ul>
+ */
 class NoCacheAzureKeyVaultOperation implements AzureKeyVaultOperation {
+    protected static final String DEFAULT_SECRET_NAME_REGEX = "^[0-9a-zA-Z-]+$";
```
Comment (edburns) [inline microprofile-config-keyvault/src/main/java/com/azure/microprofile/config/keyvault/NoCacheAzureKeyVaultOperation.java:20]: Same story here regarding `protected` vs. `private`.

**Review by edburns — CHANGES_REQUESTED** [review]
- build.yml

   Consider this comment:
   
   ```
   # In order to run tests against the real Azure services, you need to create a GitHub environment "ci" and add AZURE_CLIENT_ID / AZURE_TENANT_ID / AZURE_SUBSCRIPTION_ID as its variables to enable OIDC authentication with Azure.
   ```
   
   Can you please create setup and teardown scripts to ease the ability of someone to get started running the CI with real Azure services?

File: .github/workflows/build.yml (hunk 1 above)
Comment (majguo) [reply .github/workflows/build.yml:41]: I added them as variables as they are not secrets. I like your comment about the creation of setup and teardown scripts to set and clear these variables.

File: integration-tests/quarkus-sample/README.md (hunk 2 above)
Comment (majguo) [reply integration-tests/quarkus-sample/README.md:90]: Yes, this is expected. Will add some description in README to inform the reader that it is expected.

File: integration-tests/quarkus-sample/src/main/java/com/azure/microprofile/it/quarkus/ConfigResource.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    @Path("/value/{name}")
+    public String getConfigValue(@PathParam("name") String name) {
+        return config.getConfigValue(name).getValue();
+    }
+
+    @GET
+    @Produces(MediaType.APPLICATION_JSON)
+    @Path("/propertyNames")
+    public Set<String> getConfigPropertyNames() {
+        ConfigSource configSource = getConfigSource("AzureKeyVaultConfigSource");
```
Comment (backwind1233) [inline integration-tests/quarkus-sample/src/main/java/com/azure/microprofile/it/quarkus/ConfigResource.java:33]: Is it possible to use a const or something like AzureKeyVaultConfigSource.class.getSimpleName()?

File: microprofile-config-keyvault/src/main/java/com/azure/microprofile/config/keyvault/AzureKeyVaultConfigSource.java [hunk 5]
```diff
@@ -8,16 +8,43 @@
...
+     *  </li>
+     *  <li>
+     *      Otherwise, {@link NoCacheAzureKeyVaultOperation} will be used.
+     *      <ul>
+     *          <li>If {@code azure.keyvault.secret-name-regex} is set, then it will be used to filter secret names.</li>
+     *          <li>Otherwise, {@link NoCacheAzureKeyVaultOperation#DEFAULT_SECRET_NAME_REGEX} will be used to filter secret names.</li>
+     *      </ul>
+     *  </li>
+     * </ul>
+     */
```
Comment (backwind1233) [inline microprofile-config-keyvault/src/main/java/com/azure/microprofile/config/keyvault/AzureKeyVaultConfigSource.java:47]: ## Questions
1. Can we just call the init() once?
2. Can we log here if the value of isKeyVaultEnabled is false
image

File: .github/workflows/build.yml (hunk 1 above)
Comment (majguo) [reply .github/workflows/build.yml:41]: Addressed in [a7fb09f](https://github.com/Azure/azure-microprofile/pull/7/commits/a7fb09fb209a2a94132cb75b0d8f348db56db7c4).

File: integration-tests/quarkus-sample/README.md (hunk 2 above)
Comment (majguo) [reply integration-tests/quarkus-sample/README.md:93]: Addressed in [c261780](https://github.com/Azure/azure-microprofile/pull/7/commits/c261780ae733007f513c4226ecef7abcf55c58a7).

File: microprofile-config-keyvault/src/main/java/com/azure/microprofile/config/keyvault/CachedAzureKeyVaultOperation.java [hunk 6]
```diff
@@ -14,27 +14,51 @@
...
+/**
+ * This class is used to fetch and cache the secrets from Azure Key Vault.
+ *
+ * <ul>
+ *  <li>The cache is refreshed if it is hit but expired after the specified cacheRefreshIntervalInMs.</li>
+ *  <li>The default value of cacheRefreshIntervalInMs is 3 minutes.</li>
+ * </ul>
+ */
 class CachedAzureKeyVaultOperation implements AzureKeyVaultOperation {
+    protected static final long DEFAULT_CACHE_REFRESH_INTERVAL_IN_MS = 180000L; // 3 minutes
```
[the remark this answers is not in the document]
Comment (majguo) [reply microprofile-config-keyvault/src/main/java/com/azure/microprofile/config/keyvault/CachedAzureKeyVaultOperation.java]: Good point, fixed in [510d4a2](https://github.com/Azure/azure-microprofile/pull/7/commits/510d4a26e21be1606068969e47685f3e0cc7e6be).

File: microprofile-config-keyvault/src/main/java/com/azure/microprofile/config/keyvault/NoCacheAzureKeyVaultOperation.java (hunk 3 above)
Comment (majguo) [reply microprofile-config-keyvault/src/main/java/com/azure/microprofile/config/keyvault/NoCacheAzureKeyVaultOperation.java]: Good point, also fixed in [510d4a2](https://github.com/Azure/azure-microprofile/pull/7/commits/510d4a26e21be1606068969e47685f3e0cc7e6be).

**Comment by majguo** [comment]
> * build.yml
>   Consider this comment:
>   ```
>   # In order to run tests against the real Azure services, you need to create a GitHub environment "ci" and add AZURE_CLIENT_ID / AZURE_TENANT_ID / AZURE_SUBSCRIPTION_ID as its variables to enable OIDC authentication with Azure.
>   ```     
>     
>   Can you please create setup and teardown scripts to ease the ability of someone to get started running the CI with real Azure services?

@edburns Sure, scripts are added in commit [a7fb09f](https://github.com/Azure/azure-microprofile/pull/7/commits/a7fb09fb209a2a94132cb75b0d8f348db56db7c4).

File: integration-tests/quarkus-sample/src/main/java/com/azure/microprofile/it/quarkus/ConfigResource.java (hunk 4 above)
Comment (majguo) [reply integration-tests/quarkus-sample/src/main/java/com/azure/microprofile/it/quarkus/ConfigResource.java]: Good point, addressed in [87565b6](https://github.com/Azure/azure-microprofile/pull/7/commits/87565b6543010fe01d9d4b8acb5d5a5a66b39f8b).

File: microprofile-config-keyvault/src/main/java/com/azure/microprofile/config/keyvault/AzureKeyVaultConfigSource.java (hunk 5 above)
Comment (majguo) [reply microprofile-config-keyvault/src/main/java/com/azure/microprofile/config/keyvault/AzureKeyVaultConfigSource.java:74]: > Can we just call the init() once?

Nice comment. I re-visited the code and it turns out that `init()` can be invoked just once in the constructor. Addressed in [2a8ca00](https://github.com/Azure/azure-microprofile/pull/7/commits/2a8ca00bff508bd8e2d942c511aade1da48c3cdc).

> Can we log here if the value of isKeyVaultEnabled is false

Sure. Fixed in [7ff5098](https://github.com/Azure/azure-microprofile/pull/7/commits/7ff5098a23ed519beb857d59d3954b1ff77a0684).

**Review by edburns — APPROVED** [review]

**Pull request merged**